libparsec/crates/client/tests/unit/certif/bootstrap_workspace.rs

```diff
@@ -13,7 +16,484 @@ use crate::certif::{CertifBootstrapWorkspaceError, CertificateBasedActionOutcome
+            move |req: authenticated_cmds::latest::realm_rename::Req| {
+                RealmNameCertificate::verify_and_load(
+                    &req.realm_name_certificate,
+                    &alice.verify_key(),
+                    &alice.device_id,
+                    Some(realm_id),
+                )
+                .unwrap();
+                new_realm_certificates
+                    .lock()
+                    .unwrap()
+                    .push(req.realm_name_certificate);
+                authenticated_cmds::latest::realm_rename::Rep::Ok
+            }
+        }
+    );
+
+    let res = ops
+        .bootstrap_workspace(realm_id, &"wksp1".parse().unwrap())
+        .await
+        .unwrap();
+
+    p_assert_matches!(res, CertificateBasedActionOutcome::Uploaded { .. });
+}
+
+#[parsec_test(testbed = "minimal")]
+async fn ok_partial_bootstrap_initial_key_rotation(env: &TestbedEnv) {
+    let (env, realm_id) = env.customize_with_map(|builder| {
+        let realm_id = builder.new_realm("alice").map(|event| event.realm_id);
+        builder.certificates_storage_fetch_certificates("alice@dev1");
+        builder.rotate_key_realm(realm_id);
+
+        realm_id
+    });
+    let alice = env.local_device("alice@dev1");
+    let ops = certificates_ops_factory(&env, &alice).await;
+
+    let new_realm_certificates: Arc<Mutex<Vec<Bytes>>> = Arc::default();
+    let new_realm_initial_keys_bundle: Arc<Mutex<Option<Bytes>>> = Arc::default();
+    let new_realm_initial_keys_bundle_access: Arc<Mutex<Option<Bytes>>> = Arc::default();
+    test_register_sequence_of_send_hooks!(
+        &env.discriminant_dir,
+        // 1) Initial key rotation
+        {
+            let alice = alice.clone();
+            let new_realm_certificates = new_realm_certificates.clone();
+            let new_realm_initial_keys_bundle = new_realm_initial_keys_bundle.clone();
+            let new_realm_initial_keys_bundle_access = new_realm_initial_keys_bundle_access.clone();
+            move |req: authenticated_cmds::latest::realm_rotate_key::Req| {
+                RealmKeyRotationCertificate::verify_and_load(
+                    &req.realm_key_rotation_certificate,
+                    &alice.verify_key(),
+                    &alice.device_id,
+                    Some(realm_id),
+                )
+                .unwrap();
+                new_realm_certificates
+                    .lock()
+                    .unwrap()
+                    .push(req.realm_key_rotation_certificate);
+                *new_realm_initial_keys_bundle.lock().unwrap() = Some(req.keys_bundle);
+                let access = req
+                    .per_participant_keys_bundle_access
+                    .get(&"alice".parse().unwrap())
+                    .unwrap()
+                    .to_owned();
+                *new_realm_initial_keys_bundle_access.lock().unwrap() = Some(access);
+                authenticated_cmds::latest::realm_rotate_key::Rep::Ok
```

Comment:
should return `bad_key_index` given the initial key rotation has already been done

Reply:
Done